spinner/schema.py

```diff
@@ -0,0 +1,341 @@
+# MODELS
+# ==============================================================================
+
+
+class SpinnerMetadata(BaseModel):
+    """The metadata section of the config."""
+
+    description: str
+    version: str = Field(pattern=r"^v?\d+\.\d+(\.\d+)?$")
+    runs: int = Field(gt=0)
+    timeout: PositiveFloat | None = Field(default=None, gt=0.0)
+    retry: int = Field(default=0, ge=0)
+
+    @field_validator("retry", mode="before")
+    def validate_retry(cls, retry: int | bool) -> int:
+        if isinstance(retry, bool):
+            return 1 if retry else 0
+        return retry
+
+
+class SpinnerCommand(RootModel):
+    """A command template to execute an application."""
+
+    root: str
+
+    def __hash__(self) -> int:
+        return hash(self.template)
+
+    def template(self, env: Environment | None = None) -> Template:
+        if not env:
+            env = Environment()
+        return env.from_string(self.root)
+
+    def render(self, env: Environment | None = None, **kwargs) -> str:
+        return self.template(env).render(**kwargs)
+
+    @cached_property
+    def placeholders(self) -> set[str]:
+        return meta.find_undeclared_variables(Environment().parse(self.root))
+
+    def __str__(self) -> str:
+        return str(self.root)
+
+    def __repr__(self) -> str:
+        return repr(self.root)
+
+
+class SpinnerLambda(RootModel):
+    """A Python lambda function to parse the command output."""
+
+    root: str
+
+    @field_validator("root", mode="after")
+    def validate(cls, root: str) -> str:
+        root = root.strip()
+        try:
+            module = ast.parse(source=root)
+            assert len(module.body) == 1, "only one expression is allowed"
+            body = module.body
+            assert isinstance(body[0].value, ast.Lambda), "must be a lambda expression"
+            value = body[0].value
+            assert len(value.args.args) == 1, "lambda must receive a single argument"
+        except SyntaxError as e:
+            raise ValueError(f"syntax error: {e}") from e
+        except AssertionError as e:
+            raise ValueError(f"{e}") from e
+        return root
+
+    def __call__(self, *args, **kwargs) -> Any:
+        return eval(self.root)(*args, **kwargs)
+
+
+class SpinnerCaptureAll(BaseModel):
+    type: Literal["all"]
+    name: str
+    func: SpinnerLambda = Field(alias="lambda")
+
+    def process(self, input: str) -> tuple[str, Any]:
+        capture = None
+        return self.name, capture
+
+
+class SpinnerCaptureMatches(BaseModel):
+    type: Literal["matches"]
+    name: str
+    pattern: re.Pattern
+    func: SpinnerLambda = Field(alias="lambda")
+
+    @field_validator("pattern", mode="before")
+    def validate_pattern(cls, pattern: str) -> re.Pattern:
+        return re.compile(pattern)
+
+    def process(self, input: str) -> tuple[str, Any]:
+        capture = None
+        for line in input:
```

Comment:
This is not matching the expected lines.
I ran with the example: spinner run ./docs/examples/capture_output.yaml

And in a debugger I see that when this function is called, 'input' is always an empty string: ''

Reply:
The output from the time command is sent to stderr, not stdout:

![image](https://github.com/user-attachments/assets/6a29e57f-381b-4493-af66-59d4e6097e24)


Would it make sense to merge stderr with stdout for processing the application output?

Reply:
Testing this, I found that this function is iterating over the chars of the string.

Reply:
![image](https://github.com/user-attachments/assets/3a2004ef-c4f8-40fb-b6f1-9c13a9d23535)

Reply:
Iterate over lines to match pattern, or change how we bring the user regex here

```suggestion
        for line in input.splitlines():
```